### ISSUE TITLE
Improve geonode metadata in database

### ISSUE THREAD

**Issue by gubuntu** [kartoza/geonode#269]
_From @NyakudyaA on September 1, 2016 13:23_

# Problem

When uploading layers a user can associate the uploaded layer with already existing metadata (metadata generated by InaSAFE) using the preserve metadata tag. Alternatively, a layer can be uploaded and metadata then edited to add more information to the standard metadata generated by geonode. 

All metadata is saved in the Postgres database. When a user uploads a layer + metadata using the preserve metadata tag the metadata is stored in the database and can be properly retireved. However in instances when a user uploads a layer and then choose to edit the metadata by using the metadata UI to add the InaSAFE keywords it is impossible to retrieve fully structured metadata. 

 The metadata record is uploaded to the Postgres database  `base_resourcebase`.

An example of InaSAFE keywords to add to supplementary information: 
```
        <inasafe_keywords/>
        <inasafe>
          <population_field>
            <gco:CharacterString/>
          </population_field>
```

The metadata in ISO format looks like:
![insafe](https://user-images.githubusercontent.com/2510900/27281975-6eb0ea7e-54ee-11e7-9dbd-c27ad11b069c.png)

And when you download it it comes up as:
![safe](https://user-images.githubusercontent.com/2510900/27282124-f2613b76-54ee-11e7-92ad-627bb3f42e87.png)

# Solution

 Allow geonode UI to upload supplementary keywords and be able to download properly formatted metadata. 

_Copied from original issue: kartoza/gfdrr_oondra#30_

**Comment by gubuntu** [comment]
@nyakudya please clean up your description, it's hard to follow. This needs to go upstream to the Geonode project, which @lucernae can do

**Comment by gubuntu** [comment]
_From @lucernae on October 14, 2016 17:56_

@NyakudyaA the fastest way to do something like this is for now is to strip inasafe keywords tags and insert it to supplementalInformation using Geonode's Edit Layer UI (click the layers > Edit Layers > Edit Metadata > Supplemental Information). However this string will be escaped, so I need to do another preprocess to insert it to GeoSAFE. But with this, I would be able to easily find that InaSAFE metadata (which is in supplementalInformation and can be extracted via CSW).
The ideal solutions would be by doing modification in InaSAFE so that its keywords fully conforms GMD ISO Metadata, and I think now understand the real problem after exploring some metadata stuffs. But I need to read more about the standard.
I suggests for now we put InaSAFE keywords in supplemental information using Geonode's UI even if it will be escaped.

**Comment by gubuntu** [comment]
_From @NyakudyaA on November 25, 2016 14:10_

@lucernae  I have written database triggers to correct the special characters that are inserted into the metadata record. If we go ahead with it we will return metadata that is properly formated and this would mean some code in geosafe needs to match this. This helps us when using metadata in QGIS.  [metadata](https://github.com/kartoza/gfdrr_oondra/wiki/Ingesting-data-into--Geonode#metadata-structure)

**Comment by gubuntu** [comment]
_From @lucernae on November 28, 2016 16:1_

I would say to go with it. So, I will change the code in GeoSAFE if you already made changes to oondra.

**Comment by gubuntu** [comment]
I don't think database triggers are a good long term solution unless they are scripted into the oondra deployment repo.

**Comment by NyakudyaA** [comment]
I do not think this is still valid for all metadata. I tested with a layer from testing and the metadata is properly formatted and keywords are easy to read.

[metadata.zip](https://github.com/kartoza/geonode/files/2441858/metadata.zip)
The metadata file comes from [villages](http://testing.geonode.kartoza.com/catalogue/csw?outputschema=http%3A%2F%2Fwww.isotc211.org%2F2005%2Fgmd&service=CSW&request=GetRecordById&version=2.0.2&elementsetname=full&id=eeb24914-c28b-11e8-bb82-0220823eca32)
